code/modules/power/prototype_shield_emitter.dm

```diff
@@ -0,0 +1,208 @@
+	return TRUE
+
+/obj/machinery/power/proto_sh_emitter/process()
+	var/area/a = get_area(src)
+	var/turf/Turf = get_turf(src)
+	if(a.power_equip == FALSE && is_on == TRUE)
+		is_on = FALSE
+		update_icon_state()
+		message_admins("[src] turned off at [ADMIN_VERBOSEJMP(Turf)]")
+		log_game("[src] turned off at [AREACOORD(Turf)]")
+
+/obj/machinery/power/proto_sh_emitter/interact(mob/user)
+	var/area/a = get_area(src)
+	add_fingerprint(user)
+	if(!anchored)
+		to_chat(user, "<span class='warning'>You need to anchor the [src] first!</span>")
+		return
+	if(a.power_equip == FALSE)
+		to_chat(user, "<span class='warning'>There is no power in this area!!</span>")
+		return
+	if(locked)
+		to_chat(user, "<span class='warning'>The controls are locked!</span>")
+		return
+	if(is_on == TRUE)
+		remove_barrier(user)
+		return
+	check_map_borders(2,5,2,5)
+	if(!borders)
+		to_chat(user, "<span class='warning'>The motors whir and fail!</span>")
+		return
+	build_barrier(1,2,1,4,2,1,2,5,user)
+
+/obj/machinery/power/proto_sh_emitter/small/interact(mob/user)
+	var/area/a = get_area(src)
+	add_fingerprint(user)
+	if(!anchored)
+		to_chat(user, "<span class='warning'>You need to anchor the [src] first!</span>")
+		return
+	if(a.power_equip == FALSE)
+		to_chat(user, "<span class='warning'>There is no power in this area!!</span>")
+		return
+	if(locked)
+		to_chat(user, "<span class='warning'>The controls are locked!</span>")
+		return
+	if(is_on == TRUE)
+		remove_barrier(user)
+		return
+	check_map_borders(1,4,3,4)
+	if(!borders)
+		to_chat(user, "<span class='warning'>The motors whir and fail!</span>")
+		return
+	build_barrier(0,2,2,3,1,1,3,4,user)
+/** The vars you'll see in the proc() are referred to a mob looking north; NEx NEy refers to the North East corner x and y,
+*all the other vars works in a similar way (N = North, S = South, E = East, W = West x = x axis, y = y axis, i = internal, o = outline). This way of naming the vars
+*won't have much sense for the other directions, so always refer to the north direction when making changes as all other are already properly setup
+*This proc builds the barriers
+**/
+/obj/machinery/power/proto_sh_emitter/proc/build_barrier(SWxi,SWyi,NExi,NEyi,SWxo,SWyo,NExo,NExo,mob/user)
+	///Stores the outline of the room to generate
+	var/list/outline
+	///Stores the internal turfs of the room to generate
+	var/list/internal
+	var/turf/EmitterTurf = get_turf(src)
+	to_chat(user, "<span class='warning'>You start to turn on the [src] and the generated shields!</span>")
+	if(do_after(user, 1.5 SECONDS, target = src))
+		to_chat(user, "<span class='warning'>You turn on the [src] and the generated shields!</span>")
+		message_admins("[src] turned on at [ADMIN_VERBOSEJMP(EmitterTurf)] by [ADMIN_LOOKUPFLW(user)]")
+		log_game("[src] turned on at [AREACOORD(EmitterTurf)] by [key_name(user)]")
+		is_on = TRUE
+		update_icon_state()
+		switch(dir) //this part check the direction of the machine and create the block in front of it
```

Comment:
Don't we already have this whole field datum for this kind of bordered areas ? If there's some reason it's not applicable here this is fine.

Reply:
we have the singularity/tesla field generators, but those works by putting 4 machines in the 4 corners to create an area, and this wont work in most cases; also the main problem is that the field generators doesnt protect the turfs they are on and doesnt protect the central turfs too from the heat, allowing it to spread and destroy

Reply:
No i don't mean actualy machines i'm talking about /datum/proximity_monitor/advanced etc that handles generating zones of effects of given shape and differentiating between edge turfs and inner ones.

Reply:
it has different purposes and is way more complicated than what this needs to be.